easydigitaldownloads/easy-digital-downloads
# Modularize Javascript and Introduce Webpack

Fixes #3535 

Proposed Changes:
1. Only load relevant javascript on admin pages.

To test:

It should be ensured that all scripts listed in https://github.com/easydigitaldownloads/easy-digital-downloads/tree/test/webpack/assets/js are loaded on their necessary pages.

---

This is my proposal for updating EDD's Javascript infrastructure. I consider this "Step 1" since this implementation does not change any of the code's base architecture, and only changes how it is loaded.

The major component of this proposal is using Webpack to help bundle the Javascript. My reason for introducing Webpack is eventually I would like to see it be used for handling all of EDD's assets, and the removal of Grunt.

Webpack is currently the bundler of choice for Gutenberg (all hail) and the majority of the principles for how it is being used in this PR are based off that. Webpack works with standard Node packages which means any of the existing Grunt-wrapper packages can easily be converted to their native libraries in the future (removing dependencies).

Webpack also allows trans/compilers like Babel (which is implemented in this proposal) to use "next generation" Javascript which is not currently supported in all browsers. This is the basis for implementing this PR as it allows for easy import/export functionality between Javascript files.

This PR also introduces ESLint and the standard WordPress coding standards for Javascript (again mainly taken from Gutenberg). Linting can be run by the `npm run lint` command and automatic fixes applied with `npm run lint:fix`

---

To make this implementation easier to test and to avoid changing the development process of the repository further, minified scripts and sourcemaps have been committed to this branch. This means if a user downloads the repository they will have access to a working plugin because all scripts are included. The committed scripts are automatically in development mode and their sourcemaps will allow for easy user debugging.

In my opinion it would be more ideal in the future to remove these scripts from the repository and require developers who work directly from the repository to run the necessary build scripts. This will ensure a better developer experience (cleaner diffs, up to date scripts, and locked dependencies) while ensuring the repository remains a development-only resource. Users will still be able to download built packages from tagged releases.

---

tl;dr Everything should work exactly how it was before. This creates no functional or visual changes.
